runtime/src/main/java/com/dylibso/chicory/runtime/Instance.java

```diff
@@ -433,4 +455,512 @@ public void onExecution(Instruction instruction, long[] operands, MStack stack)
+                            if (checkName.apply(g)) {
+                                validateHostGlobalType((GlobalImport) i, g);
+                                hostGlobals[hostGlobalIdx] = g;
+                                found = true;
+                                break;
+                            }
+                        }
+                        hostGlobalIdx++;
+                        break;
+                    case MEMORY:
+                        cnt = hostImports.memoryCount();
+                        for (int j = 0; j < cnt; j++) {
+                            HostMemory m = hostImports.memory(j);
+                            if (checkName.apply(m)) {
+                                validateHostMemoryType((MemoryImport) i, m);
+                                hostMems[hostMemIdx] = m;
+                                found = true;
+                                break;
+                            }
+                        }
+                        hostMemIdx++;
+                        break;
+                    case TABLE:
+                        cnt = hostImports.tableCount();
+                        for (int j = 0; j < cnt; j++) {
+                            HostTable t = hostImports.table(j);
+                            if (checkName.apply(t)) {
+                                validateHostTableType((TableImport) i, t);
+                                hostTables[hostTableIdx] = t;
+                                found = true;
+                                break;
+                            }
+                        }
+                        hostTableIdx++;
+                        break;
+                }
+                if (!found) {
+                    if (importValidation) {
+                        throw new UnlinkableException(
+                                "unknown import, could not find host function for import number: "
+                                        + impIdx
+                                        + " named "
+                                        + name);
+                    } else {
+                        System.err.println(
+                                "Could not find host function for import number: "
+                                        + impIdx
+                                        + " named "
+                                        + name);
+                    }
+                }
+            }
+
+            var result = new HostImports(hostFuncs, hostGlobals, hostMems, hostTables);
+            return result;
+        }
+
+        private Map<String, Export> genExports(ExportSection export) {
+            var exports = new HashMap<String, Export>();
+            int cnt = export.exportCount();
+            for (int i = 0; i < cnt; i++) {
+                Export e = export.getExport(i);
+                if (exports.containsKey(e.name())) {
+                    throw new InvalidException("duplicate export name " + e.name());
+                }
```

Comment:
Shouldn't these validations be happening at the Module building level? Once the module is built they cannot change right? So won't we be re-validating this stuff everytime we make an instance from the module?

Reply:
At the moment, (most of) the validation is happening at the `Instance` building level.
We can evaluate to move some of this logic down to the `Module` building level, but let's do it in a separate PR.